tests/transaction_log_tester.ts

```diff
@@ -0,0 +1,113 @@
+    Expect(moveOperation).not.toBeNull();
+    Expect(moveOperation.source).toEqual(2);
+    Expect(moveOperation.destination).toEqual(8);
+    Expect(() => transactionLog.push(new RemoveRowOperation(5))).not.toThrow();
+    const removeOperation = operations.pop() as RemoveRowOperation;
+    Expect(removeOperation).not.toBeNull();
+    Expect(removeOperation.index).toEqual(5);
+    Expect(() => transactionLog.push(new UpdateOperation(1, 9))).not.toThrow();
+    const updateOperation = operations.pop() as UpdateOperation;
+    Expect(updateOperation).not.toBeNull();
+    Expect(updateOperation.row).toEqual(1);
+    Expect(updateOperation.column).toEqual(9);
+    listener.unlisten();
+  }
+
+  /** Tests operations with transaction in progress. */
+  @Test()
+  public testTransaction(): void {
+    const transactionLog = new TransactionLog();
+    const operations: Operation[] = [];
+    const listener = transactionLog.connect((operation: Operation) => {
+      operations.push(operation);
+    });
+    Expect(operations.length).toEqual(0);
+    Expect(() => transactionLog.beginTransaction()).not.toThrow();
+    Expect(operations.length).toEqual(0);
+    Expect(() => transactionLog.push(new AddRowOperation(25))).not.toThrow();
+    Expect(operations.length).toEqual(0);
+    Expect(() => transactionLog.beginTransaction()).not.toThrow();
+    Expect(operations.length).toEqual(0);
+    Expect(() => transactionLog.push(new MoveRowOperation(0, 10))).
+      not.toThrow();
+    Expect(operations.length).toEqual(0);
+    Expect(() => transactionLog.endTransaction()).not.toThrow();
+    Expect(operations.length).toEqual(0);
+    Expect(() => transactionLog.push(new RemoveRowOperation(7))).not.toThrow();
+    Expect(operations.length).toEqual(0);
+    Expect(() => transactionLog.push(new UpdateOperation(6, 2))).not.toThrow();
+    Expect(operations.length).toEqual(0);
+    Expect(() => transactionLog.endTransaction()).not.toThrow();
+    const transaction = operations.pop() as Transaction;
+    Expect(transaction).not.toBeNull();
+    Expect(transaction.operations.length).toEqual(4);
+    const addOperation = transaction.operations[0] as AddRowOperation;
+    Expect(addOperation).not.toBeNull();
+    Expect(addOperation.index).toEqual(25);
+    const moveOperation = transaction.operations[1] as MoveRowOperation;
+    Expect(moveOperation).not.toBeNull();
+    Expect(moveOperation.source).toEqual(0);
+    Expect(moveOperation.destination).toEqual(10);
+    const removeOperation = transaction.operations[2] as RemoveRowOperation;
+    Expect(removeOperation).not.toBeNull();
+    Expect(removeOperation.index).toEqual(7);
+    const updateOperation = transaction.operations[3] as UpdateOperation;
+    Expect(updateOperation).not.toBeNull();
+    Expect(updateOperation.row).toEqual(6);
+    Expect(updateOperation.column).toEqual(2);
+    listener.unlisten();
+  }
+
+  /** Tests empty transactions. */
+  @Test()
+  public testEmptyTransactions(): void {
```

Comment:
This isn't a bad test, but in the future, I'd make smaller tests and more of them.

Imagine that 92 fails. There are multiple reasons it could. Maybe the initial beginTransaction is bad. Or maybe it only has an error occur where a second transaction begins.

Longer tests are great as they can expose bugs that simple tests might miss, but if you aren't testing the trivial cases you're just assuming they pass and are making it harder if more complicated tests fail.